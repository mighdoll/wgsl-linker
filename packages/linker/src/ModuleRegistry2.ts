import { parseModule2, TextExport2, TextModule2 } from "./ParseModule2.js";

/** A named function to transform code fragments (e.g. by inserting parameters) */
export interface Template {
  name: string;
  apply: ApplyTemplateFn;
}
export type CodeGenFn = (
  name: string,
  params: Record<string, string>
) => string;

export interface GeneratorExport {
  name: string;
  args: string[];
  generate: CodeGenFn;
}

export type ApplyTemplateFn = (
  src: string,
  params: Record<string, any> // combination of external params and imp/exp params
) => string;

/** a single export from a module */
export type ModuleExport2 = TextModuleExport2 | GeneratorModuleExport2;

export interface TextModuleExport2 {
  module: TextModule2;
  export: TextExport2;
  kind: "text";
}

export interface GeneratorModule {
  kind: "generator";
  name: string;
  exports: GeneratorExport[];
}

export interface GeneratorModuleExport2 {
  module: GeneratorModule;
  export: GeneratorExport;
  kind: "function";
}

/** unique index for naming otherwise unnamed generator modules */
let unnamedCodeDex = 0;

/**
 * A ModuleRegistry collects exportable code fragments, code generator functions,
 * and template processors.
 *
 * The ModuleRegistry provides everything required for linkWgsl to process
 * #import statements and generate a complete wgsl shader.
 */
export class ModuleRegistry2 {
  // map from export names to a map of module names to exports
  private exports = new Map<string, ModuleExport2[]>();
  private templates = new Map<string, ApplyTemplateFn>();

  constructor(...src: string[]) {
    this.registerModules({}, ...src);
  }

  /** register modules' exports */
  registerModules(params: Record<string, any>, ...sources: string[]): void {
    sources.forEach((src) => this.registerOneModule(src, params));
  }

  /** register one module's exports  */
  registerOneModule(
    src: string,
    params: Record<string, any>,
    moduleName?: string
  ): void {
    const m = parseModule2(src, params, moduleName);
    this.addTextModule(m);
  }

  /** register a function that generates code on demand */
  registerGenerator(
    exportName: string,
    fn: CodeGenFn,
    params?: string[],
    moduleName?: string
  ): void {
    const exp: GeneratorExport = {
      name: exportName,
      args: params ?? [],
      generate: fn,
    };
    const module: GeneratorModule = {
      kind: "generator",
      name: moduleName ?? `funModule${unnamedCodeDex++}`,
      exports: [exp],
    };
    const moduleExport: GeneratorModuleExport2 = {
      module,
      export: exp,
      kind: "function",
    };
    this.addModuleExport(moduleExport);
  }

  /** register a template processor  */
  registerTemplate(...templates: Template[]): void {
    templates.forEach((t) => this.templates.set(t.name, t.apply));
  }

  /** fetch a template processor */
  getTemplate(name?: string): ApplyTemplateFn | undefined {
    return name ? this.templates.get(name) : undefined;
  }

  /** return a reference to an exported text fragment or code generator (i.e. in response to an #import request) */
  getModuleExport(
    exportName: string,
    moduleName?: string
  ): ModuleExport2 | undefined {
    const exports = this.exports.get(exportName);
    if (!exports) {
      return undefined;
    } else if (moduleName) {
      return exports.find((e) => e.module.name === moduleName);
    } else if (exports.length === 1) {
      return exports[0];
    } else {
      const moduleNames = exports.map((e) => e.module.name).join(", ");
      console.warn(
        `Multiple modules export "${exportName}". (${moduleNames}) ` +
          `Use "#import ${exportName} from <moduleName>" to select which one import`
      );
    }
  }

  private addTextModule(module: TextModule2): void {
    module.exports.forEach((e) => {
      const moduleExport: TextModuleExport2 = {
        module,
        export: e,
        kind: "text",
      };
      this.addModuleExport(moduleExport);
    });
  }

  private addModuleExport(moduleExport: ModuleExport2): void {
    const expName = exportName(moduleExport);
    const existing = this.exports.get(expName);
    if (existing) {
      existing.push(moduleExport);
    } else {
      this.exports.set(expName, [moduleExport]);
    }
  }
}

function exportName(moduleExport: ModuleExport2): string {
  if (moduleExport.kind === "text") {
    return moduleExport.export.ref.name;
  } else {
    return moduleExport.export.name;
  }
}
