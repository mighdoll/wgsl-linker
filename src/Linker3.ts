import { FnElem } from "./AbstractElems.js";
import { ModuleRegistry2 } from "./ModuleRegistry2.js";
import { TextModule2, parseModule2 } from "./ParseModule2.js";
import { BothRefs, FoundRef, recursiveRefs, traverseRefs } from "./TraverseRefs.js";
import { grouped, multiKeySet, replaceTokens3 } from "./Util.js";

export function linkWgsl3(
  src: string,
  registry: ModuleRegistry2,
  extParams: Record<string, any> = {}
): string {
  const srcModule = parseModule2(src);
  const refs = findReferences(srcModule, registry);

  const fnDecls = new Set(srcModule.fns.map(f => f.name));
  const renames = uniquify(refs, fnDecls);

  const importedText = extractTexts(refs, renames);
  return rmImports(srcModule) + "\n\n" + importedText;
}

function findReferences(
  srcModule: TextModule2,
  registry: ModuleRegistry2
): FoundRef[] {
  const visited = new Set<string>();
  const found: FoundRef[] = [];
  traverseRefs(srcModule, registry, handleRef);

  function handleRef(ref: FoundRef): boolean {
    const expImpArgs = ref.kind === "exp" ? ref.expImpArgs : [];
    const impArgs = expImpArgs.map(([_, arg]) => arg);
    const argsStr = "(" + impArgs.join(",") + ")";
    const fullName = ref.expMod.name + "." + ref.fn.name + argsStr;

    if (visited.has(fullName)) return false;

    found.push(ref);
    visited.add(fullName);
    return true;
  }
  return found;
}

type RenameMap = Map<string, Map<string, string>>;

/**
 * Calculate and @return a renaming map so that all the found top level elements
 * will have unique, non conflicting names.
 *
 * The rename map includes entries to rename references in both the exporting module
 * and the importing module.
 *
 * @param fnDecls declarations visible in the linked src so far
 */
function uniquify(refs: FoundRef[], fnDecls: Set<string>): RenameMap {
  /** number of conflicting names, used as a unique suffix for deconflicting */
  let conflicts = 0;

  /** Renames per module.
   * In the importing module the key is the 'as' name, the value is the linked name
   * In the export module the key is the export name, the value is the linked name
   */
  const renames: RenameMap = new Map();

  refs.forEach((r) => {
    const ref = r as BothRefs;

    // name proposed in the importing module (or in the local module for a support fn)
    const proposedName = ref.fromImport?.as ?? ref.fn.name;

    // name we'll actually use in the linked result
    const linkName = uniquifyName(proposedName);
    fnDecls.add(linkName);

    // record rename for this import in the exporting module
    if (linkName !== ref.fn.name) {
      multiKeySet(renames, ref.expMod.name, ref.fn.name, linkName);
    }
    //
    // record rename for this import in the importing module
    if (ref.impMod && linkName !== proposedName) {
      multiKeySet(renames, ref.impMod.name, proposedName, linkName);
    }
  });

  return renames;

  function uniquifyName(
    /** proposed name for this fn in the linked results (e.g. import as name) */
    proposedName: string
  ): string {
    let renamed = proposedName;
    if (fnDecls.has(proposedName)) {
      // create a unique name
      while (fnDecls.has(renamed)) {
        renamed = renamed + conflicts++;
      }
    }

    return renamed;
  }
}

function extractTexts(refs: FoundRef[], renames: RenameMap): string {
  return refs
    .map((r) => {
      // console.log("extracting:", r.fn.name);
      const replaces = r.kind === "exp" ? r.expImpArgs : [];
      return loadFnText(r.fn, r.expMod, renames, replaces);
    })
    .join("\n\n");
}

/** edit src to remove #import statements for clarity and also because #import
 * statements not in comments will likely cause errors in webgpu webgl parsing */
function rmImports(srcModule: TextModule2): string {
  const src = srcModule.src;
  const startEnds = srcModule.imports.flatMap((imp) => [imp.start, imp.end]);
  const slicePoints = [0, ...startEnds, src.length];
  const edits = grouped(slicePoints, 2);
  return edits.map(([start, end]) => src.slice(start, end)).join("\n");
}

function loadModuleSlice(
  mod: TextModule2,
  start: number,
  end: number,
  /** renaming from uniquificiation */
  renames: RenameMap,
  /** renaming from exp/imp params  and as name */
  replaces: [string, string][] = []
): string {
  const slice = mod.src.slice(start, end);

  const moduleRenames = renames.get(mod.name)?.entries() ?? [];

  // LATER be more precise with replacing e.g. rename for call sites, etc.
  const rewrite = Object.fromEntries([...moduleRenames, ...replaces]);
  return replaceTokens3(slice, rewrite);
}

/** extract a function from a module,
 * optionally replace export params with corresponding import arguments
 * optionally replace fn name with 'import as' name
 */
function loadFnText(
  fn: FnElem,
  mod: TextModule2,
  renames: RenameMap,
  replaces: [string, string][] = []
): string {
  const { start, end } = fn;
  return loadModuleSlice(mod, start, end, renames, replaces);
}
