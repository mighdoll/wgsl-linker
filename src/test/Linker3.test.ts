
import { expect, test } from "vitest";
import { ModuleRegistry2 } from "../ModuleRegistry2.js";
import { linkWgsl3 } from "../Linker3.js";
import { dlog } from "berry-pretty";

test("simple #import", () => {
  const myModule = `
    // #export
    fn foo() { /* fooImpl */ }
  `;

  const src = `
    // #import foo
    fn bar() {
      foo();
    }
  `;
  const registry = new ModuleRegistry2(myModule);
  const linked = linkWgsl3(src, registry);
  expect(linked).includes("fooImpl");
  expect(linked).not.includes("#import");
  expect(linked).not.includes("#export");
});

test("#import with parameter", () => {
  const myModule = `
    // #export (Elem)
    fn foo(a: Elem) { /* fooImpl */ }
  `;

  const src = `
    struct MyElem {}

    // #import foo(MyElem)
    fn bar() {
      foo();
    }
  `;
  const registry = new ModuleRegistry2(myModule);
  const linked = linkWgsl3(src, registry);
  expect(linked).includes("a: MyElem");
});

test("transitive import", () => {
  const binOpModule = `
    // #export(Elem) 
    fn binaryOp(a: Elem, b: Elem) -> Elem {
        return Elem(a.sum + b.sum); // binOpImpl
    }`;
  const reduceModule = `
    #export(work)
    fn reduceWorkgroup(index:u32) {
        let combined = binaryOp(work[index], work[index + 1u]);
    }

    #import binaryOp(MyElem)`;
  const src = `
    // #import reduceWorkgroup(myWork)
  
    fn main() {
      reduceWorkgroup(localId); // call the imported function
    }`;
  const registry = new ModuleRegistry2(binOpModule, reduceModule);
  const linked = linkWgsl3(src, registry);
  expect(linked).includes("myWork[index]");
  expect(linked).not.includes("work[");
  expect(linked).includes("binOpImpl");
});

test("#import foo as bar", () => {
  const myModule = `
    #export
    fn foo() { /* fooImpl */ }
   `;

  const src = `
    #import foo as bar

    fn main() {
      bar();
    }
   `;
  const registry = new ModuleRegistry2(myModule);
  const linked = linkWgsl3(src, registry);
  expect(linked).contains("fn bar()");
});

test("#import twice doesn't get two copies", () => {
  const module1 = `
    #export
    fn foo() { /* fooImpl */ }
  `;
  const module2 = `
    #export
    fn bar() { foo(); }

    #import foo
  `;
  const src = `
    #import bar
    #import foo

    fn main() {
      foo();
      bar();
    }
  `;
  const registry = new ModuleRegistry2(module1, module2);
  const linked = linkWgsl3(src, registry);
  const matches = linked.matchAll(/fooImpl/g);
  expect([...matches].length).toBe(1);
});

test("import transitive conflicts with main", () => {
  const module1 = `
    #export
    fn grand() {
      /* grandImpl */
    }
  `;
  const module2 = `
    #import grand
    
    #export
    fn mid() { grand(); }
  `;
  const src = `
    #import mid

    fn main() {
      mid();
    }

    fn grand() {
      /* main impl */
    }
  `;
  const registry = new ModuleRegistry2(module1, module2);
  const linked = linkWgsl3(src, registry);
  expect(linked).includes("mid() { grand0(); }");
});

test("#import twice with different names", () => {
  const module1 = `
    #export(A)
    fn foo(a:A) { /* module1 */ }
  `;
  const src = `
    #import foo(b) as bar
    #import foo(z) as zap

    fn main() {
      bar();
      zap();
    }
  `;
  const registry = new ModuleRegistry2(module1);
  const linked = linkWgsl3(src, registry);
  const matches = linked.matchAll(/module1/g);
  expect([...matches].length).toBe(2);
});

test("#import foo from zap (multiple modules)", () => {
  const module1 = `
    // #export
    fn foo() { /* module1 */ }
  `;
  const module2 = `
    // #export
    fn foo() { /* module2 */ }
  `;

  const src = `
    #import foo as baz from module2

    fn main() {
      baz();
    }
  `;

  const registry = new ModuleRegistry2();
  registry.registerOneModule(module1, "module1");
  registry.registerOneModule(module2, "module2");
  const linked = linkWgsl3(src, registry);
  expect(linked).contains("/* module2 */");
});

test("multiple exports from the same module", () => {
  const module1 = `
    #export
    fn foo() { }
    #export
    fn bar() { }
  `;
  const src = `
    #import foo
    #import bar
    fn main() {
      foo();
      bar();
    }
  `;
  const registry = new ModuleRegistry2(module1);
  const linked = linkWgsl3(src, registry);
  expect(linked).toMatchSnapshot();
});

test("#import and resolve conflicting support function", () => {
  const module1 = `
    #export
    fn foo() {
      support();
    }

    fn support() { }
  `;
  const src = `
    #import foo as bar

    fn support() { 
      bar();
    }
  `;
  const registry = new ModuleRegistry2(module1);
  const linked = linkWgsl3(src, registry);
  const origMatch = linked.matchAll(/\bsupport\b/g);
  expect([...origMatch].length).toBe(1);
  const module1Match = linked.matchAll(/\bsupport0\b/g);
  expect([...module1Match].length).toBe(2);
  const barMatch = linked.matchAll(/\bbar\b/g);
  expect([...barMatch].length).toBe(2);
});

test("#import support fn that references another import", () => {
  const src = `
    #import foo 

    fn support() { 
      foo();
    }
  `;
  const module1 = `
    #import bar

    #export
    fn foo() {
      support();
      bar();
    }

    fn support() { }
  `;
  const module2 = `
    #export
    fn bar() {
      support();
    }

    fn support() { }
  `;

  const registry = new ModuleRegistry2(module1, module2);
  const linked = linkWgsl3(src, registry);

  const origMatch = linked.matchAll(/\bsupport\b/g);
  expect([...origMatch].length).toBe(1);
  const module1Match = linked.matchAll(/\bsupport0\b/g);
  expect([...module1Match].length).toBe(2);
  const module2Match = linked.matchAll(/\bsupport1\b/g);
  expect([...module2Match].length).toBe(2);
});

test("#import support fn from two exports", () => {
  const src = `
    #import foo
    #import bar 
    fn main() {
      foo();
      bar();
    }
  `;
  const module1 = `
    #export
    fn foo() {
      support();
    }

    #export
    fn bar() {
      support();
    }

    fn support() { }
  `;

  const registry = new ModuleRegistry2(module1);
  const linked = linkWgsl3(src, registry);
  const supportMatch = linked.matchAll(/\bsupport\b/g);
  expect([...supportMatch].length).toBe(3);
});

test.only("#export importing", () => {
  const module1 = `
    #export(A, B) importing bar(A)
    fn foo(a:A, b:B) { bar(a); }

  `;
  const module2 = `
    #export(X)
    fn bar(x:X) { }
  `;
  const src = `
    #import foo(K, L)
    fn main() {
      foo(k, l);
    }
  `;
  const registry = new ModuleRegistry2(module1, module2);
  const linked = linkWgsl3(src, registry);
  console.log("linked", linked);
  expect(linked).contains("fn bar(k)");
});
//
// TODO test importing a struct constructor
test.skip("import fn with support struct constructor", () => {
  const src = `
    #import zero
    fn main() {
      let ze = zero();
    }
  `
  const module1 = `
    struct Elem {
      sum: u32;
    }

    #export 
    fn zero() -> Elem {
      return Elem(0u);
    }
  `
  const registry = new ModuleRegistry2(module1);
  const linked = linkWgsl3(src, registry);
  console.log(linked);
});


// test("import with template replace", () => {
//   const myModule = `
//     #template replacer
//     #export(threads)
//     fn foo() {
//       for (var step = 0; step < 4; step++) { //#replace 4=threads
//       }
//     }
//   `;
//   const src = `
//     #import foo(128)
//     foo();
//   `;
//   const registry = new ModuleRegistry(myModule);
//   registry.registerTemplate(replacerTemplate);
//   const linked = linkWgsl(src, registry);
//   expect(linked).includes("step < 128");
// });

// test("#import snippet w/o support functions", () => {
//   const module1 = `
//     var logVar: u32;

//     #template replacer
//     #export log(logVar, logType)
//       log(logVar, "u32"); // #replace u32=logType
//   `;

//   const src = `
//     fn foo() {
//       myVar: i32 = 1;
//       #import log(myVar, i32)
//     }
//   `;
//   const registry = new ModuleRegistry(module1);
//   registry.registerTemplate(replacerTemplate);
//   const linked = linkWgsl(src, registry);
//   expect(linked).contains('log(myVar, "i32");');
// });

// test("#import snippet w/ support functions", () => {
//   const module1 = `
//     var logVar: u32;

//     #template replacer
//     #export log(logVar, logType)
//       log(logVar);
//     #endInsert
//     fn log(myVar: u32) {} // #replace u32=logType
//   `;

//   const src = `
//     fn foo() {
//       myVar: i32 = 1;
//       #import log(myVar, i32)
//     }
//   `;
//   const registry = new ModuleRegistry(module1);
//   registry.registerTemplate(replacerTemplate);
//   const linked = linkWgsl(src, registry);
//   expect(linked).includes("log(myVar);");
//   expect(linked).includes("fn log(myVar: i32) {}");
// });

// test("resolve conflicting import support struct imports", () => {
//   const module1 = `
//     #export
//     fn foo() {
//       e: Elem = Elem(1);
//     }

//     struct Elem {
//       v: i32,
//     }

//     var <workgroup> a: array<Elem, 64>;
//   `;

//   const src = `
//      #import foo
//      #import foo as bar

//      struct Elem {
//        other: f32;
//      }

//      foo();
//      bar();
//     `;
//   const registry = new ModuleRegistry(module1);
//   const linked = linkWgsl(src, registry);
//   const origMatch = linked.matchAll(/\bElem\b/g);
//   expect([...origMatch].length).toBe(1);
//   const module1Match = linked.matchAll(/\bElem_0\b/g);
//   expect([...module1Match].length).toBe(4);
//   const module2Match = linked.matchAll(/\bElem_1\b/g);
//   expect([...module2Match].length).toBe(4);
// });

// test("#import from code generator", () => {
//   function generate(params: { name: string }): string {
//     return `fn foo() { /* ${params.name}Impl */ }`;
//   }

//   const src = `
//     #import foo(bar)

//     foo();
//   `;
//   const registry = new ModuleRegistry();
//   registry.registerGenerator("foo", generate as CodeGenFn, ["name"]);
//   const linked = linkWgsl(src, registry);
//   expect(linked).contains("barImpl");
// });

// test("#import as with code generator", () => {
//   function generate(params: { name: string }): string {
//     return `fn foo() { /* ${params.name}Impl */ }`;
//   }

//   const src = `
//     #import foo(bar) as baz

//     baz();
//   `;
//   const registry = new ModuleRegistry();
//   registry.registerGenerator("foo", generate as CodeGenFn, ["name"]);
//   const linked = linkWgsl(src, registry);
//   expect(linked).contains("fn baz()");
// });

// test("#import code generator snippet with support", () => {
//   function generate(params: { name: string; logType: string }): TextInsert {
//     return {
//       src: `log(${params.name});`,
//       rootSrc: `fn log(logVar: ${params.logType}) {}`,
//     };
//   }

//   const src = `
//     fn foo() {
//       let bar: i32 = 1
//       #import log(bar, i32)
//     }
//   `;
//   const registry = new ModuleRegistry();
//   registry.registerGenerator("log", generate as CodeGenFn, ["name", "logType"]);
//   const linked = linkWgsl(src, registry);
//   expect(linked).contains("log(bar);");
//   expect(linked).contains("fn log(logVar: i32) {}");
// });

// test("external param applied to template", () => {
//   const module1 = `
//     #template replacer
//     #export(threads)
//     fn foo() {
//       for (var step = 0; step < 4; step++) { //#replace 4=threads
//       }
//     }
//   `;
//   const src = `
//     #import foo(workgroupThreads)
//     foo();
//   `;
//   const registry = new ModuleRegistry(module1);
//   registry.registerTemplate(replacerTemplate);
//   const params = { workgroupThreads: 128 };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).includes("step < 128");
// });

// test("#if", () => {
//   const src = `
//     #if foo
//       foo();
//     #endif
//     #if bar
//       bar();
//     #endif
//   `;
//   const registry = new ModuleRegistry();
//   const params = { foo: true, bar: false };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).toMatchSnapshot();
// });

// test("nested #if", () => {
//   const src = `
//     #if foo
//       #if bar
//         foo();
//       #endif
//       #if zap
//         zap();
//       #endif
//     #endif
//   `;
//   const registry = new ModuleRegistry();
//   const params = { foo: true, bar: false, zap: true };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).toMatchSnapshot();
// });

// test("#endExport", () => {
//   const module1 = `
//     struct Foo {
//     //  #export field
//       sum: u32;
//     // #endExport
//     }`;
//   const src = `
//     struct MyStruct {
//       // #import field
//     }`;
//   const registry = new ModuleRegistry(module1);
//   const linked = linkWgsl(src, registry);
//   expect(linked).toMatchSnapshot();
// });

// test("#template in src", () => {
//   const src = `
//     #template replacer
//     fn main() {
//       for (var step = 0; step < 4; step++) { //#replace 4=threads
//       }
//     }
//   `;
//   const registry = new ModuleRegistry();
//   registry.registerTemplate(replacerTemplate);
//   const params = { threads: 128 };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).includes("step < 128");
// });

// test("#if !foo", () => {
//   const src = `
//     #if !foo
//       bar();
//     #endif
//   `
//   const registry = new ModuleRegistry();
//   const params = { };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).includes("bar();");
// });

// test("#if #else", () => {
//   const src = `
//     #if foo
//       foo();
//     #else
//       bar();
//     #endif
//   `
//   const registry = new ModuleRegistry();
//   const params = { };
//   const linked = linkWgsl(src, registry, params);
//   expect(linked).not.includes("foo();");
//   expect(linked).includes("bar();");
// });
