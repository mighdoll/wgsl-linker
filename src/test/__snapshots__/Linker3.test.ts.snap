// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`multiple exports from the same module 1`] = `
"
    

    

    fn main() {
      foo();
      bar();
    }
  

fn foo() { }

fn bar() { }"
`;
